### PR TITLE
SONARJAVA-1986 Branch only to corresponding catch block

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-frontend/src/test/files/se/ConditionAlwaysTrueOrFalseCheck.java [hunk 1]
```diff
@@ -1613,12 +1613,13 @@ void testAfterAddAssignment(int y) {
...
 }
 
-abstract class TryCatchCFG {
+public class TryCatchCFG {
 
   private Object monitor;
   private boolean shutdown;
 
-  abstract void doSomething();
+  private void doSomething() {
```
Comment (m-g-sonar) [inline java-frontend/src/test/files/se/ConditionAlwaysTrueOrFalseCheck.java]: This revert the change from SONARJAVA-1985, to get back to master state.

File: java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java [hunk 2]
```diff
@@ -191,20 +193,33 @@ public void computedFrom(List<SymbolicValue> symbolicValues) {
...
     }
 
     @Override
     public String toString() {
-      return super.toString() + "_" + exceptionType.fullyQualifiedName() + "!";
+      return super.toString() + "_" + (exceptionType == null ? "!unknownException" : exceptionType.fullyQualifiedName()) + "!";
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java]: MINOR This class overrides "equals()" and should therefore also override "hashCode()". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1206)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* MAJOR 2 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [CFG.java](https://github.com/SonarSource/sonar-java/blob/4db2dc6db08a6f25b1bf974b997fc91623799d2f/java-frontend/src/main/java/org/sonar/java/cfg/CFG.java): This file has 848 lines, which is greater than 750 authorized. Split it into smaller files. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00104)
1. MAJOR [ExplodedGraphWalker.java](https://github.com/SonarSource/sonar-java/blob/4db2dc6db08a6f25b1bf974b997fc91623799d2f/java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java): This file has 877 lines, which is greater than 750 authorized. Split it into smaller files. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00104)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 3]
```diff
@@ -582,39 +587,111 @@ private void executeMethodInvocation(MethodInvocationTree mit) {
...
+      // Enqueue exceptional paths from exceptional yields
+      methodInvokedBehavior.exceptionalPathYields().flatMap(yield -> yield.statesAfterInvocation(
+        invocationArguments,
+        invocationTypes,
+        programState,
+        () -> thrownExceptionsByExceptionType.computeIfAbsent(yield.exceptionType, constraintManager::createExceptionalSymbolicValue))
+      ).forEach(ps -> enqueueExceptionalPaths(ps.clearStack(), (SymbolicValue.ExceptionalSymbolicValue) ps.peekValue()));
 
+      // Enqueue additional exceptional paths corresponding to unchecked exceptions, for instance OutOfMemoryError
+      enqueueUncheckedExceptionalPaths();
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: factorize that before the if.

File: java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 4]
```diff
@@ -582,39 +587,111 @@ private void executeMethodInvocation(MethodInvocationTree mit) {
       List<SymbolicValue> invocationArguments = invocationArguments(unstack.values);
       List<Type> invocationTypes = mit.arguments().stream().map(ExpressionTree::symbolType).collect(Collectors.toList());
 
-      methodInvokedBehavior.exceptionalPathYields()
-        .flatMap(yield -> yield.statesAfterInvocation(invocationArguments, invocationTypes, programState, () -> resultValue).stream())
-        .forEach(this::enqueueExceptionalPaths);
+      Map<Type, SymbolicValue.ExceptionalSymbolicValue> thrownExceptionsByExceptionType = new HashMap<>();
+
+      // Enqueue exceptional paths from exceptional yields
+      methodInvokedBehavior.exceptionalPathYields().flatMap(yield -> yield.statesAfterInvocation(
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: please indent correctly and put new lines like the happyPathYields : flat map should be on next line.

File: java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 5]
```diff
@@ -582,39 +587,111 @@ private void executeMethodInvocation(MethodInvocationTree mit) {
...
+      // Enqueue happy paths
+      programState = handleSpecialMethods(programState.stackValue(resultValue), mit, methodSymbol);
       checkerDispatcher.executeCheckPostStatement(mit);
       clearStack(mit);
     }
   }
 
-  private void enqueueExceptionalPaths(ProgramState ps) {
-    node.programPoint.block.exceptions().forEach(b -> enqueue(new ExplodedGraph.ProgramPoint(b, 0), ps.clearStack(), !b.isCatchBlock()));
+  private ProgramState handleSpecialMethods(ProgramState ps, MethodInvocationTree mit, Symbol methodSymbol) {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: name is not great here (in fact this method should probably not be here, but in the checks?). I would not have the 3rd parameter and extract symbol from mit locally (for readability)

File: java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 6]
```diff
@@ -582,39 +587,111 @@ private void executeMethodInvocation(MethodInvocationTree mit) {
...
+    if (isNonNullMethod(methodSymbol)) {
+      return ps.addConstraint(ps.peekValue(), ObjectConstraint.NOT_NULL);
+    } else if (OBJECT_WAIT_MATCHER.matches(mit)) {
+      return ps.resetFieldValues(constraintManager);
+    }
+    return ps;
+  }
+
+  private void enqueueThrownExceptionalPaths(Symbol symbol) {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: I think this one could rely on MethodInvocationTree : that would allow to not declare methodSymbol local variable.

File: java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 7]
```diff
@@ -582,39 +587,111 @@ private void executeMethodInvocation(MethodInvocationTree mit) {
...
+    // store the exception as exit value in case of method exit in next block
+    ProgramState newPs = ps.clearStack().stackValue(exceptionSV);
+    newPs.storeExitValue();
+
+    node.programPoint.block.successors().stream()
+      .filter(CFG.Block::isMethodExitBlock)
+      .forEach(b -> enqueue(new ExplodedGraph.ProgramPoint(b, 0), newPs, true));
+  }
+
+  private static boolean isValidCatchBlock(CFG.Block catchBlock, @Nullable Type thrownType) {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: isCaughtByBlock ? and swapping the parameters ?

File: java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 8]
```diff
@@ -582,39 +587,111 @@ private void executeMethodInvocation(MethodInvocationTree mit) {
...
+      .forEach(b -> enqueue(new ExplodedGraph.ProgramPoint(b, 0), ps));
+
+    // use other exceptional blocks, i.e. finally block and exit blocks
+    exceptionBlocks.stream()
+      .filter(CFG.Block.IS_CATCH_BLOCK.negate())
+      .forEach(b -> enqueue(new ExplodedGraph.ProgramPoint(b, 0), ps, true));
+
+    // store the exception as exit value in case of method exit in next block
+    ProgramState newPs = ps.clearStack().stackValue(exceptionSV);
+    newPs.storeExitValue();
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: This does not handle properly the case where you have a return in finally (and where you have not a return in the finally in fact)

File: java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java (hunk 6 above)
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: Not convinced. The `methodSymbol` is still required to get the declaration and for `methodCanNotBeOverriden`, line 592.

Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: This one can be inlined.
